packages/native/src/main.udl

```diff
@@ -12,123 +12,73 @@ enum Error {
   "ManifestError",
   "FileReadError",
   "ResolverError",
-  "PluginError"
+  "PluginError",
+  "RuntimeError"
 };
 
-interface Uri {
+interface FFIUri {
+  constructor([ByRef] string authority, [ByRef] string path, [ByRef] string uri);
+
+  [Name=from_string]
   constructor([ByRef] string uri);
+
+  string to_string();
 };
 
 interface FFIInvoker {
-  [Throws=Error]
-  sequence<u8> invoke_raw(Uri uri, [ByRef] string method, sequence<u8>? args, string? env);
+  [Throws=FFIError]
+  sequence<u8> invoke_raw(FFIUri uri, [ByRef] string method, sequence<u8>? args, sequence<u8>? env, FFIUriResolutionContext? resolution_context);
 
-  [Throws=Error]
-  sequence<Uri> get_implementations(Uri uri);
+  [Throws=FFIError]
+  sequence<FFIUri> get_implementations(FFIUri uri);
 
-  record<DOMString, sequence<Uri>>? get_interfaces();
+  record<DOMString, sequence<FFIUri>>? get_interfaces();
 };
 
-interface FFIWrapper {
-  [Throws=Error]
-  sequence<u8> invoke(FFIInvoker invoker, Uri uri, [ByRef] string method, sequence<u8>? args, string? env);
+callback interface FFIAbortHandler {
+  void abort(string msg);
 };
 
-interface FFIWrapPackage { };
+callback interface FFIWrapper {
+  sequence<u8> invoke([ByRef] string method, sequence<u8>? args, sequence<u8>? env, FFIInvoker invoker, FFIAbortHandler? abort_handler);
+};
+
+callback interface FFIWrapPackage {
+  FFIWrapper create_wrapper();
+};
 
 interface FFIWasmWrapper {
   constructor(sequence<u8> wasm_module);
-};
 
-callback interface FFIPluginModule {
-  sequence<u8> wrap_invoke(string method_name, sequence<u8> params, string? env, FFIInvoker invoker);
+  [Throws=FFIError]
+  sequence<u8> invoke([ByRef] string method, sequence<u8>? args, sequence<u8>? env, FFIInvoker invoker, FFIAbortHandler? abort_handler);
 };
 
 callback interface FFIUriResolver {
-  FFIUriPackageOrWrapper wrap_try_resolve_uri(Uri uri, FFIInvoker invoker);
+  FFIUriPackageOrWrapper try_resolve_uri(FFIUri uri, FFIInvoker invoker);
```

Comment:
Is this missing resolution context arg?

Reply:
Indeed it was missing it. I am now passing it but it isn't that simple, due to Uniffi constraints and our current `UriResolver` trait definition.

Refs can't be passed across the boundary; but smart pointers can. The fact that our `try_resolve_uri` method takes a `&mut UriResolutionContext` instead of `Arc<Mutex<UriResolutionContext>>` makes it impossible for the resolution context to behave as it normally does when an externally-implemented `FFIUriResolver` is found in the resolution chain. Ideally, we'd create a structure that wraps the `&mut` ref; but that would make the struct expose lifetime constraints and we wouldn't be able to expose it through FFI

I drafted a compromise, without modifying our trait definitions, which works in the following way: when an external resolver is found, the current resolution_context is cloned; the external resolution function is called, and when it completes, the original resolution_context is updated from the external function result.

The simpler option is to make `try_resolve_ur`i take an `Arc<Mutex<UriResolutionContext>>`. This would be a general change just to satisfy a Uniffi constraint; which seems incorrect to me but maybe in this case the benefits outweigh the cons.